### PR TITLE
Split API and WebApp routes

### PR DESCRIPTION
**Pull request by saneery**
https://github.com/poanetwork/blockscout/issues/2364

## Motivation
To launch API endpoints separately from the web application. 

There are three groups of endpoints:
* Read-only API
* Web app
* Write endpoints

## Changelog

* Split API's and WebApp's endpoints into separate routers
* Save old endpoints for backward compatibility
* Update tests
* Add envs:
   * `DISABLE_WEBAPP=true`
   * `DISABLE_READ_API =true`
   * `DISABLE_INDEXER=true`
   * `DISABLE_WRITE_API =true`
   * `WEBAPP_URL=http://host/path`
   * `API_URL=http://host/path`

### CODE REVIEW

**Comment by saneery** [comment]
> @saneery
> 
> If I set `export DISABLE_WEBAPP=true` before launch
> 
> 1. Indexer still works - I see new blocks. It shouldn't

Maybe I should add new ENV var `DISABLE_INDEXER` ?

**Comment by vbaranov** [comment]
>@vbaranov Maybe I should add new ENV var DISABLE_INDEXER ?

@saneery  yes, it makes sense in order to increase manageability.

**Comment by saneery** [comment]
> @saneery I disabled indexer + webapp. It seems ETH_RPC doesn't work to me:
> 
> ```
> viktor$ curl -H "Content-type: application/json" -H "Content-type: application/json" -X POST --data '{"jsonrpc":"2.0","method":"eth_blockNumber","id":1}' localhost:4000/api/eth_rpc
> {"jsonrpc":"2.0","error": "id is a required field","id": 0}
> 
> viktor$ curl -H "Content-Type: application/json" -X POST --data '{"jsonrpc":"2.0","method":"eth_getBalance","params":["0x053229e7866226d8ffb0e6ee916d9a11b806c678", "latest"],"id":1}' localhost:4000/api/eth_rpc
> {"jsonrpc":"2.0","error": "Balance not found","id": 1}
> ```
> 
> Does it work for you?

@vbaranov 

First method `eth_blockNumber` isn't supported in BS. API supports only two methods `eth_getBalance` and `eth_getLogs` https://blockscout.com/poa/core/eth_rpc_api_docs.

In the second method, I think the address doesn't exist in your DB.
I had an address `0xb6695f5c2e3F5EFf8036b5F5F3A9d83A5310e51E` and got this

```bash
$ curl -H "content-type: application/json" -X POST --data '{"jsonrpc":"2.0","method":"eth_getBalance","id":1, "params": ["0xb6695f5c2e3F5EFf8036b5F5F3A9d83A5310e51E", "latest"]}' localhost:4000/api/eth_rpc
{"jsonrpc":"2.0","result":"0xffffe90685f830800","id":1}
```

**Comment by vbaranov** [comment]
@saneery please fix merging conflicts

**Comment by saneery** [comment]
> @saneery please fix merging conflicts

@vbaranov done

**Comment by vbaranov** [comment]
@saneery gettext test is broken

**Comment by saneery** [comment]
@vbaranov 
1. I made market history catalog switched off when indexer is disabled. 

2. I think, when API and Webapp will hosts on different machines there will be different `BLOCKSCOUT_HOST` and `NETWORK_PATH` vars, and API will contain own vars.
I guess we need a new var that will contain URL to Webapp, something like a `WEBAPP_URL`

**Comment by vbaranov** [comment]
@saneery 
>I think, when API and Webapp will hosts on different machines there will be different BLOCKSCOUT_HOST and NETWORK_PATH vars, and API will contain own vars.
I guess we need a new var that will contain URL to Webapp, something like a WEBAPP_URL

Yes, let's add it. Please ping me when it will be ready.

**Comment by saneery** [comment]
@vbaranov ping :smile: 
I added new var `WEBAPP_URL` it must be like a `http://host/path`

**Comment by saneery** [comment]
@vbaranov I've added new variables `DISABLE_UPDATE_ENDPOINTS` and `API_URL` could you check this?

**Comment by saneery** [comment]
> If I set
> 
> ```
> DISABLE_API=true
> DISABLE_UPDATE_ENDPOINTS=false
> ```
> 
> All API methods are available, but with this set of parameters only APIs, which update DB should be available

@vbaranov Endpoints in phoenix's router are set on compile-time, so`DISABLE_API, DISABLE_WEBAPP, DISABLE_UPDATE_ENDPOINTS` works only on compile-time, maybe you didn't compile BS with this params? 
Use this commands
```
DISABLE_API=true DISABLE_UPDATE_ENDPOINTS=false mix compile --force
DISABLE_API=true DISABLE_UPDATE_ENDPOINTS=false mix phx.server
```

I get this error then try call API method
```bash
$ curl -H "content-type: application/json" -sSL -D - -X POST --data '{"jsonrpc":"2.0","method":"eth_getBalance","id":1, "params": ["0xb6695f5c2e3F5EFf8036b5F5F3A9d83A5310e51E", "latest"]}' localhost:4000/api/eth_rpc

HTTP/1.1 404 Not Found
```

And this when try call writing method
```bash
$ curl -H "content-type: application/json" -sSL -D - -X POST --data '{}' localhost:4000/api/v1/verified_smart_contracts

HTTP/1.1 422 Unprocessable Entity
{"error":"address_hash is invalid"}%  
```

**Comment by saneery** [comment]
> @saneery
> 
> If I don't set any of new env vars, I expect that app should work in a full mode, but when I go to any item in APIs menu the application shows me 404 page instead of (eth) api docs or graphiql

@vbaranov :thinking: I don't have 404 on documentation pages

**Comment by saneery** [comment]
@vbaranov I don't know how to reproduce your issue

**Comment by vbaranov** [comment]
@saneery yes, I should revoke my last review. I tested all possible combinations of parameters. Seems, it works. Please resolve merging conflicts. And you didn't update `docs/env-variables.md` with the finalized set of ENV vars.

**Comment by vbaranov** [comment]
And it is odd that all tests on the real node failed.

**Comment by saneery** [comment]
@vbaranov @pasqu4le @ayrat555 I've repaired test, could you check the PR?

File: apps/block_scout_web/test/block_scout_web/controllers/api/v1/decompiled_smart_contract_controller_test.exs [hunk 1]
```diff
@@ -3,6 +3,7 @@ defmodule BlockScoutWeb.API.V1.DecompiledControllerTest do
 
   alias Explorer.Repo
   alias Explorer.Chain.{Address, DecompiledSmartContract}
+  alias BlockScoutWeb.ApiRouter.Helpers
```
Comment (pasqu4le) [inline apps/block_scout_web/test/block_scout_web/controllers/api/v1/decompiled_smart_contract_controller_test.exs]: I see you had to add a lot of these `alias`es, would't it be easier to add an `import` for them in `BlockScoutWeb.ConnCase` as for the other `Helpers` modules?
If the problem is namespace clashing, a sensible qualified `alias` still seems worth it.

**Pull request merged**